maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/BinaryDownloader.java

```diff
@@ -33,42 +34,69 @@ private File download() {
         File downloaded = checkIfItIsAlreadyDownloaded();
         if (downloaded == null) {
             markerFileHandler.createMarkerFile();
-
+            URL redirectedMavenDistribution = checkForRedirect(mavenDistribution);
+            long expectedSize = 0;
             for (int i = 0; i < 3; i++) {
                 try {
-                    downloaded = runDownloadExecution(mavenDistribution, target).await();
-                } catch (ExecutionException ee) {
+                    if (expectedSize == 0) {
```

Comment:
I don't understand what are you doing here, but if you run this only only first iteration, then why not to move it before the loop.

Reply:
In case of IOException caused by the HttpURLConnection the block will be run again. Moving it before the loop would mean to have another loop for this case